06.bowling_object/frame.rb

```diff
@@ -0,0 +1,45 @@
+# frozen_string_literal: true
+
+require_relative './shot'
+
+class Frame
+  TOTAL_FRAMES = 10
+
+  attr_reader :scores, :shots
+
+  def initialize(shot_scores, frame_number)
+    @frame_number = frame_number
+    @shots = shot_scores.map { |score| Shot.new(score) }
+  end
+
+  def calculate_point(frames)
+    @shots.sum(&:point) + calculate_additional_point(frames)
+  end
+
+  private
+
+  def calculate_additional_point(frames)
+    return 0 if @frame_number == TOTAL_FRAMES
+
+    shot_count = calculate_shot_count
+    frames[@frame_number..]
+      .flat_map(&:shots)
+      .take(shot_count)
+      .sum(&:point)
+  end
+
+  def calculate_shot_count
+    n = 0
+    n = 2 if strike?
+    n = 1 if spare?
+    n
```

Comment:
変数への再代入は基本的に避けた方がいいので、変更前の方が良かったですね。
もしくはこんなふうに書くならアリかなーと思います。

```suggestion
    return 2 if strike?
    return 1 if spare?
    0
```

Reply:
レビューありがとうございます！

[こちら](https://github.com/taco-nantai/ruby-practices/pull/10#discussion_r1456638812)で頂いたレビューで、再代入を使った書き方への修正を提案していただいたので、今回こちらのコードも再代入を使った書き方に修正したのですが、再代入を使うべき状況と使ってはいけない状況があるということでしょうか。また、そうであった場合、その基準はどのようなものになるでしょうか。
再代入の使い方についてもう少し説明をいただきたいです。お手数をおかけしますがよろしくお願いします🙇‍♂️🙇‍♂️

Reply:
> こちらで頂いたレビューで、再代入を使った書き方への修正を提案していただいたので、今回こちらのコードも再代入を使った書き方に修正したのですが、再代入を使うべき状況と使ってはいけない状況があるということでしょうか。また、そうであった場合、その基準はどのようなものになるでしょうか。

やっぱりそこが気になりますよね😅

まず、原則として変数への再代入はなるべく避けた方がいいです。
なので、20行目で再代入するのも文句なしのベストとは言えません。

ただ、再代入を避けようとするとこんなコードになります。

```ruby
n =
  if frame_number == Frame::TOTAL_FRAMES
    copied_scores.size
  else
    copied_scores[0] == 'X' ? 1 : 2
  end
```

もしくは三項演算子をネストさせる方法もありますが、可読性が最悪なのでこれはやめた方がいいですね。

```ruby
n = frame_number == Frame::TOTAL_FRAMES ? copied_scores.size : copied_scores[0] == 'X' ? 1 : 2
```

というわけで再代入させないコードと再代入を許すコードを比べると、後者の方がスマートだったのであえて後者を選びました。
（再代入NGはあくまで原則であって、絶対NGではない）

```ruby
# こっちの方がスマート（個人的には）
n = copied_all_scores[0] == 'X' ? 1 : 2
n = copied_all_scores.size if frame_number == Frame::TOTAL_FRAMES
```

ただし、変数への代入ではなくメソッドの戻り値の場合はまた考え方が別です。
たとえば、以下のように`n`の値を別メソッドで算出するロジックにしたケースを考えてみます。

```ruby
n = calc_shift_value(copied_scores, frame_number)
```

これなら次のような実装にすることができます。

```ruby
def calc_shift_value(scores, frame_number)
  return scores.size if frame_number == Frame::TOTAL_FRAMES

  scores[0] == 'X' ? 1 : 2
end
```

メソッドの場合は早期リターンというテクニックが使える点が、メソッド無しで`n`に代入する場合と違う点ですね。

もちろん、早期リターンを使わないやり方もありです（好み）。

```ruby
def calc_shift_value(scores, frame_number)
  if frame_number == Frame::TOTAL_FRAMES
    scores.size
  else
    scores[0] == 'X' ? 1 : 2
  end
end
```

よって、`calculate_shot_count` は変数の再代入を使わない方法（かつ早期リターンを使う方法）をおすすめしたのでした。

こんな感じで僕の意図が伝わったでしょうか？

Reply:
回答ありがとうございます！
再代入を許容する基準がよくわかりました！
今後はなるべく再代入を使わないようにしつつ、早期リターンを活用していきたいと思います。
ありがとうございました🙇‍♂️🙇‍♂️